packages/internal/src/debug.js

```diff
@@ -1,41 +1,108 @@
 // @jessie-check
 
+/**
+ * @import {OptPrefixPath} from './types-internal.js';
+ */
+
+// TODO This is top level static mutable state. Why did it pass at-jessie-check?
 let debugInstance = 1;
 
+const logDisabled = (..._args) => {};
+logDisabled.sub = _ => logDisabled;
+harden(logDisabled);
+
 /**
- * @param {string} name
+ * @typedef {typeof logDisabled} TraceLogger
+ */
+
+/**
+ * @param {OptPrefixPath} optPrefixPath
+ * @param {number} rootInstanceNum
  * @param {boolean | 'verbose'} enable
+ * @returns {TraceLogger}
  */
-export const makeTracer = (name, enable = true) => {
-  debugInstance += 1;
+const makeSubTracer = (optPrefixPath, rootInstanceNum, enable) => {
+  // DO NOT ASSIGN to the parameter variables above, since they are captured
+  // by closures below.
+  const makeKey = () => {
+    // As much as possible, postpone anything expensive to here, since this is
+    // only called when a tracer or sub-tracer function is called. Not when it
+    // is made.
+    let prefix;
+    // instead, make a local assignable copy
+    let optPath = optPrefixPath;
+    while (optPath !== undefined) {
+      const [step, optRestSteps] = optPath;
+      if (prefix === undefined) {
+        prefix = step;
+      } else {
+        prefix = `${step}.${prefix}`;
+      }
+      optPath = optRestSteps;
+    }
+    return `----- ${prefix}.${rootInstanceNum} `;
+  };
+  const sub = prefixStep =>
```

Comment:
The biggest remaining change I plan to do in this PR is to add essentially the same optional `enable` parameter as found on `makeTracer`. This will enable tracing to be turned on or off for any subtree, except as overridden (shadowed?) by deeper subtrees.

```suggestion
  const sub = (prefixStep, subEnable = enable) =>
```

Reply:
Done at #11857 . Resolving.